File: daily/155. 最小栈.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    
};

var stack=[],min=[];
/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    if(min.length==0 || min[min.length-1]>=x)
        min.push(x);
    stack.push(x);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if(stack[stack.length-1]==min[min.length-1])
        min.pop();
    stack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return stack[stack.length-1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return min[min.length-1];
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: daily/C++版本/155. 最小栈.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    stack<int> stack,flag;
    MinStack() {
        
    }
    
    void push(int x) {
        if(flag.empty() || flag.top()>=x)
            flag.push(x);
        stack.push(x);
    }
    
    void pop() {
        if(stack.top()==flag.top())
            flag.pop();
        stack.pop();
    }
    
    int top() {
        return stack.top();
    }
    
    int getMin() {
        return flag.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */